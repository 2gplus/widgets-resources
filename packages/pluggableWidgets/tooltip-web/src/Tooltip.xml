<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.tooltip.Tooltip"
        needsEntityContext="true"
        pluginWidget="true"
        offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Tooltip</name>
    <description/>
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/tooltip</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="trigger" type="widgets">
                    <caption>Trigger</caption>
                    <description/>
                </property>
                <property key="render" type="enumeration" defaultValue="text">
                    <caption>Render method</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="text">Text</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="content" type="widgets" required="false">
                    <caption>Content</caption>
                    <description/>
                </property>

                <property key="tooltipString" type="textTemplate" required="false">
                    <caption>Tooltip</caption>
                    <description/>
                </property>

                <property key="position" type="enumeration" defaultValue="top">
                    <caption>Position</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="topStart">Top Start</enumerationValue>
                        <enumerationValue key="top">Top</enumerationValue>
                        <enumerationValue key="topEnd">Top End</enumerationValue>
                        <enumerationValue key="leftStart">Left Start</enumerationValue>
                        <enumerationValue key="left">Left</enumerationValue>
                        <enumerationValue key="leftEnd">Left End</enumerationValue>
                        <enumerationValue key="bottomStart">Bottom Start</enumerationValue>
                        <enumerationValue key="bottom">Bottom</enumerationValue>
                        <enumerationValue key="bottomEnd">Bottom End</enumerationValue>
                        <enumerationValue key="rightStart">Right Start</enumerationValue>
                        <enumerationValue key="right">Right</enumerationValue>
                        <enumerationValue key="rightEnd">Right End</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="openOn" type="enumeration" defaultValue="hover">
                    <caption>Open on</caption>
                    <description>When to show the tooltip. On a mobile device the 'On hover', will be triggered on
                        touch.
                    </description>
                    <enumerationValues>
                        <enumerationValue key="click">Click</enumerationValue>
                        <enumerationValue key="hover">Hover</enumerationValue>
                        <enumerationValue key="hoverFocus">Hover &amp; focus</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>