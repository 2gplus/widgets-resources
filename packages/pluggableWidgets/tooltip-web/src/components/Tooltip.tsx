import { createElement, CSSProperties, ReactElement, ReactNode, useCallback, useRef, useState } from "react";
import { useOnClickOutside } from "@mendix/piw-utils-internal/components/web";
import { OpenOnEnum, RenderEnum } from "../../typings/TooltipProps";
import { Placement } from "@popperjs/core/lib/enums";
import { usePopper } from "react-popper";
import classNames from "classnames";

export interface TooltipProps {
    name?: string;
    class?: string;
    style?: CSSProperties;
    tabIndex?: number;
    trigger: ReactNode;
    render: RenderEnum;
    content: ReactNode;
    tooltipString?: string;
    position: Placement;
    openOn: OpenOnEnum;
    preview?: boolean;
}

export const Tooltip = (props: TooltipProps): ReactElement => {
    const { trigger, tooltipString, content, openOn, position, preview, render } = props;
    const componentReference = useRef<HTMLDivElement | null>(null);
    const [showTooltip, setShowTooltip] = useState(preview ?? false);
    const [referenceElement, setReferenceElement] = useState<HTMLDivElement | null>(null);
    const [popperElement, setPopperElement] = useState<HTMLDivElement | null>(null);
    const [arrowElement, setArrowElement] = useState<HTMLDivElement | null>(null);
    const { styles, attributes } = usePopper(referenceElement, popperElement, {
        placement: position,
        modifiers: [
            {
                name: "arrow",
                options: {
                    element: arrowElement,
                    padding: 5
                }
            },
            // TODO should this property be dynamic? Ask Chris!
            { name: "offset", options: { offset: [0, 8] } }
        ]
    });

    useOnClickOutside([componentReference], () => setShowTooltip(false));

    const onHover = useCallback((show: boolean) => setShowTooltip(show), []);

    const onClick = useCallback((show: boolean) => setShowTooltip(!show), []);

    const renderTrigger = (): ReactElement => {
        let eventContainer;
        switch (openOn) {
            case "click":
                eventContainer = {
                    onClick: () => onClick(showTooltip)
                };
                break;
            case "hover":
                eventContainer = {
                    onMouseEnter: () => onHover(true),
                    onMouseLeave: () => onHover(false)
                };
                break;
            case "hoverFocus":
                eventContainer = {
                    onMouseEnter: () => onHover(true),
                    onMouseLeave: () => onHover(false),
                    onFocus: () => onHover(true),
                    onBlur: () => onHover(false)
                };
                break;
        }
        return (
            <div
                className={"widget-tooltip-trigger"}
                data-testid={"trigger"}
                ref={setReferenceElement}
                {...eventContainer}
            >
                {trigger}
            </div>
        );
    };

    const renderContent = (): ReactNode => {
        return showTooltip ? (
            <div
                className={"widget-tooltip"}
                ref={setPopperElement}
                style={styles.popper}
                {...attributes.popper}
                role={"tooltip"}
            >
                {render === "text" ? tooltipString : content}
                <div
                    className={"widget-tooltip-arrow"}
                    ref={setArrowElement}
                    style={styles.arrow}
                    {...attributes.arrow}
                />
            </div>
        ) : null;
    };
    return (
        <div className={classNames(props.class, "widget-tooltip-container")} ref={componentReference}>
            {renderTrigger()}
            {renderContent()}
        </div>
    );
};
