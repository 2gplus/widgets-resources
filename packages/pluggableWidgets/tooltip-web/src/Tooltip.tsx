import { ReactNode, createElement } from "react";
import { Tooltip as DisplayTooltip } from "./components/Tooltip";
import { TooltipContainerProps } from "../typings/TooltipProps";
import { translatePosition } from "./utils";
import "./ui/Tooltip.scss";

export default function Tooltip(props: TooltipContainerProps): ReactNode {
    return (
        <DisplayTooltip
            position={translatePosition(props.position)}
            content={props.content}
            class={props.class}
            name={props.name}
            openOn={props.openOn}
            render={props.render}
            trigger={props.trigger}
            tooltipString={props.tooltipString?.value}
            style={props.style}
            tabIndex={props.tabIndex}
        />
    );
}
