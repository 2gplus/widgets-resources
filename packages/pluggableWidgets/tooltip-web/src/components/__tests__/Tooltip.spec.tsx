import { createElement } from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";

import { Tooltip, TooltipProps } from "../Tooltip";

describe("Tooltip", () => {
    let defaultTooltipProps: TooltipProps;

    beforeEach(() => {
        defaultTooltipProps = {
            content: undefined,
            tooltipString: "Tooltip text",
            position: "right",
            openOn: "click",
            trigger: "",
            render: "text",
            name: "tooltip"
        };
    });

    it("should render", () => {
        const { asFragment } = render(<Tooltip {...defaultTooltipProps} />);
        const triggerElement = screen.getByTestId("trigger");
        act(() => {
            fireEvent.click(triggerElement);
        });
        expect(asFragment()).toMatchSnapshot();
    });
    it("should render and open tooltip onMouseEnter and close onMouseLeave", () => {
        render(<Tooltip {...defaultTooltipProps} openOn={"hover"} />);
        const triggerElement = screen.getByTestId("trigger");
        act(() => {
            fireEvent.mouseEnter(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        act(() => {
            fireEvent.mouseLeave(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();

        act(() => {
            fireEvent.focus(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();
        act(() => {
            fireEvent.click(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("should render and open tooltip onClick", () => {
        render(<Tooltip {...defaultTooltipProps} openOn={"click"} />);
        const triggerElement = screen.getByTestId("trigger");
        act(() => {
            fireEvent.click(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        act(() => {
            fireEvent.click(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("should render and open tooltip onFocus and close onBlur", () => {
        defaultTooltipProps.openOn = "hoverFocus";
        render(<Tooltip {...defaultTooltipProps} />);
        const triggerElement = screen.getByTestId("trigger");
        act(() => {
            fireEvent.focus(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        act(() => {
            fireEvent.blur(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();

        act(() => {
            fireEvent.mouseEnter(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        act(() => {
            fireEvent.mouseLeave(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("should render text content if the tooltipString is passed", () => {
        const { asFragment } = render(<Tooltip {...defaultTooltipProps} render={"text"} />);
        act(() => {
            fireEvent.click(screen.getByTestId("trigger"));
        });
        expect(screen.queryByRole("tooltip")).toHaveTextContent(defaultTooltipProps.tooltipString as string);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render HTML if the content is passed", () => {
        act(() => {
            render(
                <Tooltip
                    {...defaultTooltipProps}
                    render={"custom"}
                    content={<div data-testid={"content"}>Simple Tooltip</div>}
                    tooltipString={undefined}
                />
            );
        });
        act(() => {
            fireEvent.click(screen.getByTestId("trigger"));
        });
        expect(screen.queryByTestId("content")).toBeInTheDocument();
        expect(screen.queryByTestId("content")).toHaveTextContent("Simple Tooltip");
    });

    it("should close onOutsideClick if tooltip is visible", () => {
        render(<Tooltip {...defaultTooltipProps} openOn={"click"} />);
        const triggerElement = screen.getByTestId("trigger");
        act(() => {
            fireEvent.click(triggerElement);
        });
        expect(screen.queryByRole("tooltip")).toBeInTheDocument();

        act(() => {
            fireEvent.mouseDown(document);
        });
        expect(screen.queryByRole("tooltip")).toBeNull();
    });
});
