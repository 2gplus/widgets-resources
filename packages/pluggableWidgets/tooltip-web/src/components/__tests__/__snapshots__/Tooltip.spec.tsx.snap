// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render 1`] = `
<DocumentFragment>
  <div
    class="widget-tooltip-container"
  >
    <div
      class="widget-tooltip-trigger"
      data-testid="trigger"
    />
    <div
      class="widget-tooltip"
      role="tooltip"
      style="position: absolute; left: 0px; top: 0px;"
    >
      Tooltip text
      <div
        class="widget-tooltip-arrow"
        style="position: absolute;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip should render text content if the tooltipString is passed 1`] = `
<DocumentFragment>
  <div
    class="widget-tooltip-container"
  >
    <div
      class="widget-tooltip-trigger"
      data-testid="trigger"
    />
    <div
      class="widget-tooltip"
      role="tooltip"
      style="position: absolute; left: 0px; top: 0px;"
    >
      Tooltip text
      <div
        class="widget-tooltip-arrow"
        style="position: absolute;"
      />
    </div>
  </div>
</DocumentFragment>
`;
